src/libtorch.cc

```diff
@@ -552,11 +582,48 @@ ModelInstanceState::Create(
 
 ModelInstanceState::ModelInstanceState(
     ModelState* model_state, TRITONBACKEND_ModelInstance* triton_model_instance)
+#ifdef TRITON_ENABLE_GPU
+    : BackendModelInstance(model_state, triton_model_instance),
+      model_state_(model_state), device_(torch::kCPU), is_dict_input_(false),
+      pytorch_stream_(at::cuda::getDefaultCUDAStream())
+#else
     : BackendModelInstance(model_state, triton_model_instance),
       model_state_(model_state), device_(torch::kCPU), is_dict_input_(false)
+#endif
 {
   if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
     device_ = torch::Device(torch::kCUDA, DeviceId());
+#ifdef TRITON_ENABLE_GPU
+    pytorch_stream_ =
+        at::cuda::getStreamFromPool(false /* is_high_priority */, DeviceId());
```

Comment:
Does this guarantee to not returning stream that being used by other instances? I don't see places to release the stream back to the pool.

And what is the difference between using this stream v.s. the one we created in backend utils (`stream_`), would it achieve the same if we wrap `stream_` into `at::cuda::CUDAStream` if the library exposes API to do so?

Reply:
I will ask the pytorch team to make sure but I think as long as the object is not destructed it will not be returned to the pool. I think the release will happen when the object is destructed.

> And what is the difference between using this stream v.s. the one we created in backend utils (stream_), would it achieve the same if we wrap stream_ into at::cuda::CUDAStream if the library exposes API to do so?

I think at::cuda::CUDAStream is just a lightweight wrapper around the cudaStream. There is now way to create a PyTorch cudaStream from our own cudaStream. That is why I have used PyTorch's stream as our stream for input/output collectors.

Reply:
Switched to getstreamfromexternal https://pytorch.org/cppdocs/api/function_namespacec10_1_1cuda_1a399b2870dd4f6e4f9424666762080332.html#function-c10-cuda-getstreamfromexternal.

This API is more recently added. We construct a PyTorch stream using the model instance stream.